### ISSUE TITLE
How to train the instance segmentation, when the mask is not continuous

### ISSUE THREAD

**Issue by tswc** [ultralytics/ultralytics#21697]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

As mentioned, for coco dataset each line represent a object mask.
But for my data, the mask may not be continous. So it will be some polylines, they belong to the same obj, but  are seperated. 

Is there any config like "group ID" to solve this problem?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @tswc, thank you for your interest in Ultralytics 🚀! This is an automated response to help you get started while an Ultralytics engineer reviews your issue and will assist you shortly.

For general guidance, we recommend a visit to the Docs where you can find many Python and CLI usage examples and where many of the most common questions may already be answered:
- Docs: https://docs.ultralytics.com/
- Python: https://docs.ultralytics.com/usage/python/
- CLI: https://docs.ultralytics.com/usage/cli/

If this is a 🐛 Bug Report, please provide a minimum reproducible example to help us debug it:
- MRE: https://docs.ultralytics.com/help/minimum-reproducible-example/

If this is a custom training ❓ Question (e.g., dataset formatting, masks/polygons behavior), please include:
- A few representative image+label samples (screenshots welcome)
- Exact data.yaml and model config used
- A short snippet showing how you launch training
- Relevant training logs
And please verify you are following our Tips for Best Training Results:
- Training tips: https://docs.ultralytics.com/guides/model-training-tips/

Join the Ultralytics community where it suits you best:
- Discord (real-time chat) 🎧: https://discord.com/invite/ultralytics
- Discourse (in-depth Q&A): https://community.ultralytics.com/
- Subreddit: https://www.reddit.com/r/Ultralytics/

## Upgrade

Upgrade to the latest ultralytics package including all requirements in a Python>=3.8 environment with PyTorch>=1.8 to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```
- Requirements: https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including CUDA/CUDNN, Python and PyTorch preinstalled):

- Notebooks with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- Google Cloud Deep Learning VM. See GCP Quickstart Guide: https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/
- Amazon Deep Learning AMI. See AWS Quickstart Guide: https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/
- Docker Image. See Docker Quickstart Guide: https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/ <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all Ultralytics CI tests are currently passing. CI tests verify correct operation of all YOLO Modes and Tasks on macOS, Windows, and Ubuntu every 24 hours and on every commit.

To help us assist you faster regarding non-contiguous masks for instance segmentation, please share:
1.A tiny sample of your labels showing how multiple disjoint polygons for a single object are represented
2.Your current annotation format (COCO JSON snippets or YOLO label masks) and any “grouping” fields you’re using
3.How you convert your annotations to the Ultralytics/COCO format
4.The exact command/config you run for training and the model (e.g., YOLO11-seg)
5.Screenshots of predictions vs. ground truth if available

Thanks for the detailed info in advance 🙏

**Comment by Y-T-G** [comment]
Ultralytics doesn't support masks with holes currently. You will have to stretch a small invisible line.

**Comment by tswc** [comment]
> Ultralytics doesn't support masks with holes currently. You will have to stretch a small invisible line.

OK， thanks

**Comment by glenn-jocher** [comment]
You’re welcome—if you want to automate this, the merge_multi_segment utility and masks2segments will connect multi-part contours into a single polygon, and convert_coco with use_segments=True does this during COCO→YOLO conversion; see the merge_multi_segment reference, masks2segments docs, and the COCO to YOLO conversion guide.

**Comment by OBEsensia** [comment]
@tswc At E-sensIA we are actualy having "uncontinuous" segementation masks for our trainings and inferences working well as @Y-T-G said the trick is to generate one polygone that merges all parts of the mask with a line that is 0 pixel thick (kind of round trip closing the mask).
As @glenn-jocher proposed it converters are great for that. In any case you need something to do more atomic stuff, one function (used in converters) is available to do that here at Ultralytics [merge_multi_segment](https://docs.ultralytics.com/reference/data/converter/#ultralytics.data.converter.merge_multi_segment). Well used it is pretty efficient so I hope it will help.

**Comment by glenn-jocher** [comment]
Great tip—if your source is binary masks, you can also convert them straight to YOLO segmentation labels per class using convert_segment_masks_to_yolo_seg; see Converting Segmentation Masks to YOLO Format in the docs for details.
